futures-util/src/stream/futures_unordered/mod.rs

```diff
@@ -566,15 +573,36 @@ impl<Fut> FuturesUnordered<Fut> {
 
 impl<Fut> Drop for FuturesUnordered<Fut> {
     fn drop(&mut self) {
-        // When a `FuturesUnordered` is dropped we want to drop all futures
-        // associated with it. At the same time though there may be tons of
-        // wakers flying around which contain `Task<Fut>` references
-        // inside them. We'll let those naturally get deallocated.
-        while !self.head_all.get_mut().is_null() {
-            let head = *self.head_all.get_mut();
-            let task = unsafe { self.unlink(head) };
-            self.release_task(task);
+        // Before the strong reference to the queue is dropped we need all
+        // futures to be dropped. See note at the bottom of this method.
+        //
+        // This ensures we drop all futures even in the case of one drop
+        // panicking and unwinding.
+        //
+        // If a second future panics, that will trigger an abort from panicking
+        // while panicking.
+        struct DropGuard<'a, Fut>(&'a mut FuturesUnordered<Fut>);
+        impl<Fut> DropGuard<'_, Fut> {
+            fn drop_futures(&mut self) {
+                // When a `FuturesUnordered` is dropped we want to drop all futures
+                // associated with it. At the same time though there may be tons of
+                // wakers flying around which contain `Task<Fut>` references
+                // inside them. We'll let those naturally get deallocated.
+                while !self.0.head_all.get_mut().is_null() {
+                    let head = *self.0.head_all.get_mut();
+                    let task = unsafe { self.0.unlink(head) };
+                    self.0.release_task(task);
+                }
+            }
+        }
+        impl<Fut> Drop for DropGuard<'_, Fut> {
+            fn drop(&mut self) {
+                self.drop_futures();
+            }
         }
+        let mut guard = DropGuard(self);
+        guard.drop_futures();
+        mem::forget(guard); // no need to check head_all again
```

Comment:
This was added to make sure all futures are dropped if one panics. Although it looks like tasks have a check for this in their drop that aborts. At least with this, leaking and aborts may be avoided if only one future panics.